pkg/graveler/graveler.go

```diff
@@ -1879,7 +1880,116 @@ func (g *KVGraveler) Revert(ctx context.Context, repositoryID RepositoryID, bran
 }
 
 func (g *KVGraveler) Merge(ctx context.Context, repositoryID RepositoryID, destination BranchID, source Ref, commitParams CommitParams, strategy string) (CommitID, error) {
-	return g.db.Merge(ctx, repositoryID, destination, source, commitParams, strategy)
+	var (
+		preRunID         string
+		storageNamespace StorageNamespace
+		commit           Commit
+		commitID         CommitID
+	)
+
+	repo, err := g.RefManager.GetRepository(ctx, repositoryID)
+	if err != nil {
+		return "", err
+	}
+	storageNamespace = repo.StorageNamespace
+
+	// No retries on any failure during the merge. If the branch changed, it's either that commit is in progress, commit occurred,
+	// or some other branch changing operation. If commit is in-progress, then staging area wasn't empty after we checked so not retrying is ok.
+	// If another commit/merge succeeded, then the user should decide whether to retry the merge.
+	err = g.RefManager.BranchUpdate(ctx, repositoryID, destination, func(branch *Branch) (*Branch, error) {
+		empty, err := g.stagingEmpty(ctx, repositoryID, repo, branch)
+		if err != nil {
+			return nil, fmt.Errorf("check if staging empty: %w", err)
+		}
+		if !empty {
+			return nil, fmt.Errorf("%s: %w", destination, ErrDirtyBranch)
+		}
+		fromCommit, toCommit, baseCommit, err := g.getCommitsForMerge(ctx, repositoryID, source, Ref(destination))
+		if err != nil {
+			return nil, err
+		}
+		g.log.WithFields(logging.Fields{
```

Comment:
We should log the operation at the Merge level so that we get a logging of all attempts regardless of failures